src/main/java/appeng/fluids/helper/DualityFluidInterface.java

```diff
@@ -0,0 +1,632 @@
+	{
+		if( capabilityClass == CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY )
+		{
+			return (T) this.storage;
+		}
+		else if( capabilityClass == Capabilities.STORAGE_MONITORABLE_ACCESSOR )
+		{
+			return (T) this.accessor;
+		}
+		return null;
+	}
+
+	private boolean hasConfig()
+	{
+		return this.hasConfig;
+	}
+
+	private void readConfig()
+	{
+		final boolean had = this.hasWorkToDo();
+
+		for( int x = 0; x < NUMBER_OF_TANKS; x++ )
+		{
+			this.configStacks[x] = null;
+			ItemStack is = this.config.getStackInSlot( x );
+			if( !is.isEmpty() )
+			{
+				IFluidHandlerItem fh = is.getCapability( CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null );
+				if( fh != null )
+				{
+					FluidStack fs = fh.drain( Fluid.BUCKET_VOLUME, false );
+					if( fs != null && fs.amount > 0 )
+					{
+						this.configStacks[x] = AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ).createStack( fs );
+					}
+				}
+			}
+			this.updatePlan( x );
+		}
+
+		final boolean has = this.hasWorkToDo();
+
+		if( had != has )
+		{
+			try
+			{
+				if( has )
+				{
+					this.gridProxy.getTick().alertDevice( this.gridProxy.getNode() );
+				}
+				else
+				{
+					this.gridProxy.getTick().sleepDevice( this.gridProxy.getNode() );
+				}
+			}
+			catch( final GridAccessException e )
+			{
+				// :P
+			}
+		}
+
+		this.notifyNeighbors();
+	}
+
+	private boolean updateStorage()
+	{
+		boolean didSomething = false;
+		for( int x = 0; x < NUMBER_OF_TANKS; x++ )
+		{
+			if( this.requireWork[x] != null )
+			{
+				didSomething = this.usePlan( x ) || didSomething;
+			}
+		}
+		return didSomething;
+	}
+
+	private boolean hasWorkToDo()
+	{
+		for( final IAEFluidStack requiredWork : this.requireWork )
+		{
+			if( requiredWork != null )
+			{
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private int getTankSlot( final IFluidHandler inventory )
+	{
+		for( int i = 0; i < NUMBER_OF_TANKS; ++i )
+		{
+			if( this.tanks[i] == inventory )
+			{
+				return i;
+			}
+		}
+		throw new IndexOutOfBoundsException();
+	}
+
+	private void updatePlan( final int slot )
+	{
+		final IAEFluidStack req = this.configStacks[slot];
+		final FluidStack stored = this.tanks[slot].drain( TANK_CAPACITY, false );
+
+		if( req == null && ( stored != null && stored.amount > 0 ) )
+		{
+			final IAEFluidStack work = AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ).createStack( stored );
+			this.requireWork[slot] = work.setStackSize( -work.getStackSize() );
+			return;
+		}
+		else if( req != null )
+		{
+			if( stored == null || stored.amount == 0 ) // need to add stuff!
+			{
+				this.requireWork[slot] = req.copy();
+				return;
+			}
+			else if( req.getFluid().equals( stored.getFluid() ) ) // same type ( qty different? )!
+			{
+				if( stored.amount < TANK_CAPACITY )
+				{
+					this.requireWork[slot] = req.copy();
+					this.requireWork[slot].setStackSize( TANK_CAPACITY - stored.amount );
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Cast one of the operands of this subtraction operation to a "long". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2184)